File: 3398.smallest-substring-with-identical-characters-i.py
#
# @lc app=leetcode id=3398 lang=python3
#
# [3398] Smallest Substring With Identical Characters I
#

# @lc code=start
class Solution:
    def minLength(self, s: str, numOps: int) -> int:
        n = len(s)
        
        def check(length):
            if length == 1:
                # Check cost to make 010101...
                cost1 = 0
                for i in range(n):
                    if i % 2 == 0:
                        if s[i] != '0': cost1 += 1
                    else:
                        if s[i] != '1': cost1 += 1
                
                # Check cost to make 101010...
                cost2 = 0
                for i in range(n):
                    if i % 2 == 0:
                        if s[i] != '1': cost2 += 1
                    else:
                        if s[i] != '0': cost2 += 1
                
                return min(cost1, cost2) <= numOps
            
            # For length >= 2, we calculate flips needed for each consecutive block
            # A block of size L needs L // (length + 1) flips to ensure no sub-block > length
            ops_needed = 0
            i = 0
            while i < n:
                j = i
                while j < n and s[j] == s[i]:
                    j += 1
                
                run_length = j - i
                ops_needed += run_length // (length + 1)
                i = j
            
            return ops_needed <= numOps

        left, right = 1, n
        ans = n
        
        while left <= right:
            mid = (left + right) // 2
            if check(mid):
                ans = mid
                right = mid - 1
            else:
                left = mid + 1
                
        return ans
# @lc code=end

File: 3398.smallest-substring-with-identical-characters-i.cpp
#include <bits/stdc++.h>
using namespace std;

// @lc app=leetcode id=3398 lang=cpp
//
// [3398] Smallest Substring With Identical Characters I
//

// @lc code=start
class Solution {
public:
    int minLength(string s, int numOps) {
        int n = (int)s.size();

        auto feasible = [&](int L) -> bool {
            const int INF = 1e9;
            vector<vector<int>> dp(2, vector<int>(L + 1, INF));

            // i = 0 initialization
            for (int c = 0; c <= 1; c++) {
                int cost = ((s[0] - '0') != c);
                dp[c][1] = min(dp[c][1], cost);
            }

            for (int i = 1; i < n; i++) {
                vector<vector<int>> ndp(2, vector<int>(L + 1, INF));
                for (int prev = 0; prev <= 1; prev++) {
                    for (int len = 1; len <= L; len++) {
                        int cur = dp[prev][len];
                        if (cur >= INF) continue;
                        for (int c = 0; c <= 1; c++) {
                            int nlen = (c == prev) ? (len + 1) : 1;
                            if (nlen > L) continue;
                            int add = ((s[i] - '0') != c);
                            ndp[c][nlen] = min(ndp[c][nlen], cur + add);
                        }
                    }
                }
                dp.swap(ndp);
            }

            int best = INT_MAX;
            for (int c = 0; c <= 1; c++)
                for (int len = 1; len <= L; len++)
                    best = min(best, dp[c][len]);
            return best <= numOps;
        };

        int lo = 1, hi = n;
        while (lo < hi) {
            int mid = lo + (hi - lo) / 2;
            if (feasible(mid)) hi = mid;
            else lo = mid + 1;
        }
        return lo;
    }
};
// @lc code=end
